sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java

```diff
@@ -382,4 +346,218 @@ ItemDeserializer getItemDeserializer() {
+                                    itemClass, true)));
+                        }
+                    }
+                    return Mono.error(unwrappedException);
+                }));
+    }
+
+    public void setRequestHeaders(CosmosItemRequestOptions requestOptions) {
+        this.cosmosItemRequestOptionsAccessor.setHeader(requestOptions, Constants.IS_CLIENT_ENCRYPTED_HEADER, "true");
+        this.cosmosItemRequestOptionsAccessor.setHeader(requestOptions, Constants.INTENDED_COLLECTION_RID_HEADER, this.encryptionProcessor.getContainerRid());
+    }
+
+    public void setRequestHeaders(CosmosQueryRequestOptions requestOptions) {
+        this.cosmosQueryRequestOptionsAccessor.setHeader(requestOptions, Constants.IS_CLIENT_ENCRYPTED_HEADER, "true");
+        this.cosmosQueryRequestOptionsAccessor.setHeader(requestOptions, Constants.INTENDED_COLLECTION_RID_HEADER, this.encryptionProcessor.getContainerRid());
+    }
+
+    private <T> CosmosPagedFlux<T> queryItemsHelper(SqlQuerySpec sqlQuerySpec,
+                                                    CosmosQueryRequestOptions options,
+                                                    Class<T> classType,
+                                                    boolean isRetry) {
+        setRequestHeaders(options);
+        CosmosQueryRequestOptions finalOptions = options;
+        Flux<FeedResponse<T>>  tFlux = CosmosBridgeInternal.queryItemsInternal(container, sqlQuerySpec, options,
+            new Transformer<T>() {
+                @Override
+                public Function<CosmosPagedFluxOptions, Flux<FeedResponse<T>>> transform(Function<CosmosPagedFluxOptions, Flux<FeedResponse<JsonNode>>> func) {
+                    return queryDecryptionTransformer(classType, func);
+                }
+            }).byPage().onErrorResume(exception -> {
+            final Throwable unwrappedException = Exceptions.unwrap(exception);
+            if (unwrappedException instanceof CosmosException) {
+                final CosmosException cosmosException = (CosmosException) unwrappedException;
+                if (!isRetry && cosmosException.getStatusCode() == HttpConstants.StatusCodes.BADREQUEST &&
+                    cosmosException.getResponseHeaders().get(HttpConstants.HttpHeaders.SUB_STATUS)
+                        .equals(Constants.INCORRECT_CONTAINER_RID_SUB_STATUS)) {
+                    this.encryptionProcessor.getIsEncryptionSettingsInitDone().set(false);
+                    return this.encryptionProcessor.initializeEncryptionSettingsAsync(true).thenMany(
+                        (CosmosPagedFlux.defer(() -> queryItemsHelper(sqlQuerySpec,finalOptions, classType, true).byPage())));
+                }
+            }
+            return Mono.error(unwrappedException);
+        });
+
+
+        CosmosQueryRequestOptions finalOptions3 = options;
+        return UtilBridgeInternal.createCosmosPagedFlux(pagedFluxOptions -> {
```

Comment:
Wrapping FeedResponse to create CosmosPagedFlux twice as we dont have onErrorResume on CosmosPagedFlux .

Reply:
what's the impact on the Tracer? do we log events twice?

Reply:
Yes you are right , it will trace twice , removed it , other side effect is client telemetry is captured twice, handled the scenario in new iteration